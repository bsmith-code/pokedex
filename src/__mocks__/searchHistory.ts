export const mockSearchHistory = Array.from({ length: 20 }).map(
  (_, idx) => `Search ${idx}`
)
