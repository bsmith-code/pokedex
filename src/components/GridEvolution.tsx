import { IEvolutionChain } from 'pokeapi-typescript'

import { Box, Grid, ListItemButton, styled, Typography } from '@mui/material'

import { usePokemonContext } from 'context/PokemonContext'

import { flattenChain, getIdFromUrl, getPokemonSprite } from 'utils'

interface IProps {
  isFetching: boolean
  evolutions?: IEvolutionChain
}

const StyledGridItem = styled(ListItemButton)(({ theme }) => ({
  flexDirection: 'column',
  borderRadius: '4px',
  color: '#fff',
  padding: theme.spacing(4),
  backgroundColor: theme.palette.secondary.main,
  '&:hover': {
    backgroundColor: '#6704c9'
  }
}))

export const GridEvolution = ({ evolutions, isFetching }: IProps) => {
  const { handleChangePokemon } = usePokemonContext()
  const flattenedChain = flattenChain(evolutions?.chain)

  return !isFetching ? (
    <Box>
      <Typography variant="h6" sx={{ mb: 2 }}>
        Evolutions
      </Typography>
      {flattenedChain?.length ? (
        <Grid container spacing={2}>
          {flattenedChain.map(({ url, name }) => {
            const id = getIdFromUrl(url)
            return (
              <Grid item xs={12} sm={6} md={6} lg={4} key={`evolution-${name}`}>
                <StyledGridItem onClick={() => handleChangePokemon(name)}>
                  <img
                    alt={name}
                    src={getPokemonSprite(id)}
                    width="120px"
                    height="120px"
                    style={{ objectFit: 'contain' }}
                  />
                  <Typography variant="subtitle2">{name}</Typography>
                </StyledGridItem>
              </Grid>
            )
          })}
        </Grid>
      ) : (
        <Typography>No available evolutions</Typography>
      )}
    </Box>
  ) : (
    <Typography>Loading...</Typography>
  )
}
