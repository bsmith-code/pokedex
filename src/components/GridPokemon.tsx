import {
  Box,
  Grid,
  ListItemButton,
  Pagination,
  styled,
  Typography
} from '@mui/material'

import { usePokemonContext } from 'context/PokemonContext'

import { getIdFromUrl, getPokemonSprite } from 'utils'

import {
  TESTID_GRID_ERROR,
  TESTID_GRID_LOADING,
  TESTID_GRID_POKEMON
} from 'constants/testIds'

const StyledGridItem = styled(ListItemButton)`
  flex-direction: column;
  border-radius: 4px;
  border: 1px solid #efefef;
`

export const GridPokemon = () => {
  const {
    pokemon,
    currentPage,
    paginatedPokemon,
    isErrorPokemon,
    isFetchingPokemon,
    handleChangePage,
    handleChangePokemon
  } = usePokemonContext()

  if (isFetchingPokemon) {
    return <div data-testid={TESTID_GRID_LOADING}>Loading...</div>
  }

  if (!pokemon.length || isErrorPokemon) {
    return <div data-testid={TESTID_GRID_ERROR}>Pokemon unavailable.</div>
  }

  return (
    <Box mt={4} data-testid={TESTID_GRID_POKEMON}>
      <Grid container spacing={2}>
        {paginatedPokemon.map(({ url, name }) => {
          const imageSrc = getPokemonSprite(getIdFromUrl(url))

          return (
            <Grid
              item
              xs={12}
              sm={6}
              md={4}
              lg={3}
              key={`pokemon-grid-${name}`}
            >
              <StyledGridItem onClick={() => handleChangePokemon(name)}>
                <img
                  alt={name}
                  src={imageSrc}
                  width="120px"
                  height="120px"
                  style={{ objectFit: 'contain' }}
                />
                <Typography>{name}</Typography>
              </StyledGridItem>
            </Grid>
          )
        })}
      </Grid>
      <Box mt={6} display="flex" justifyContent="center">
        <Pagination
          page={currentPage}
          count={Math.floor(pokemon.length / 20)}
          onChange={(_, page) => handleChangePage(page)}
        />
      </Box>
    </Box>
  )
}
