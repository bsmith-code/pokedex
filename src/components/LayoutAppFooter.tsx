import { Toolbar, Typography } from '@mui/material'

const LayoutAppFooter = () => {
  const year = new Date().getFullYear()

  return (
    <Toolbar
      variant="dense"
      component="footer"
      sx={{ bgcolor: 'primary.main', justifyContent: 'center' }}
    >
      <Typography color="#fff">
        &copy; Brian Matthew Smith, {year}. All Rights Reserved.
      </Typography>
    </Toolbar>
  )
}

export default LayoutAppFooter
