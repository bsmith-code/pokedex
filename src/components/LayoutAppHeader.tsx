import { useNavigate } from 'react-router-dom'

import { IconButton, Toolbar } from '@mui/material'

const LayoutAppHeader = () => {
  const navigate = useNavigate()

  const handleLogoClick = () => {
    navigate('/')
  }

  return (
    <Toolbar component="header" sx={{ bgcolor: 'primary.main' }}>
      <IconButton onClick={handleLogoClick}>
        <img
          alt="PrizePicks"
          src="/assets/logo.svg"
          width="140px"
          height="auto"
        />
      </IconButton>
    </Toolbar>
  )
}

export default LayoutAppHeader
