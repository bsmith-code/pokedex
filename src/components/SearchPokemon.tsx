import { INamedApiResource } from 'pokeapi-typescript'

import {
  Autocomplete,
  createFilterOptions,
  ListItemButton,
  TextField
} from '@mui/material'

import { useChangePokemon } from 'hooks/useChangePokemon'

interface IProps {
  pokemon: INamedApiResource<void>[]
}

export const SearchPokemon = ({ pokemon }: IProps) => {
  const { handleChangePokemon } = useChangePokemon()

  return (
    <Autocomplete
      filterOptions={createFilterOptions({ limit: 20 })}
      onChange={(_, value) => handleChangePokemon(value?.name ?? '')}
      options={pokemon}
      renderInput={params => <TextField {...params} label="Search Pokemon" />}
      getOptionLabel={({ name }) => name}
      renderOption={(props, { name }) => (
        <ListItemButton
          {...props}
          component="li"
          key={`pokemon-list-${name}`}
          data-testid={`pokemon-${name}`}
        >
          {name}
        </ListItemButton>
      )}
    />
  )
}
