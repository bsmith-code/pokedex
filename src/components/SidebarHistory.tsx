import { Box, List, ListItemButton, Paper, Typography } from '@mui/material'

import { usePokemonContext } from 'context/PokemonContext'

import { TESTID_SEARCH_EMPTY, TESTID_SEARCH_HISTORY } from 'constants/testIds'

export const SidebarHistory = () => {
  const { searchHistory, handleChangePokemon } = usePokemonContext()

  return (
    <Box
      px={4}
      py={3}
      flexShrink={1}
      flexBasis={300}
      component={Paper}
      bgcolor="grey.200"
      sx={{ '@media (max-width: 767px)': { order: 2, flexGrow: 1 } }}
    >
      <Typography variant="subtitle2">Search History</Typography>

      {searchHistory.length ? (
        <List data-testid={TESTID_SEARCH_HISTORY}>
          {searchHistory.map(name => (
            <ListItemButton
              component="li"
              key={`search-${name}`}
              onClick={() => handleChangePokemon(name)}
            >
              {name}
            </ListItemButton>
          ))}
        </List>
      ) : (
        <Typography data-testid={TESTID_SEARCH_EMPTY} sx={{ mt: 2 }}>
          No previous searches.
        </Typography>
      )}
    </Box>
  )
}
