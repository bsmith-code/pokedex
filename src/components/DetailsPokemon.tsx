import { useNavigate } from 'react-router-dom'

import ArrowBackIcon from '@mui/icons-material/ArrowBack'
import {
  Box,
  Card,
  CardContent,
  CardHeader,
  IconButton,
  Typography
} from '@mui/material'

import { usePokemonDetails } from 'hooks/usePokemonDetails'

import { GridEvolution } from 'components/GridEvolution'

import {
  TESTID_DETAILS_ERROR,
  TESTID_DETAILS_LOADING,
  TESTID_DETAILS_POKEMON,
  TESTID_DETAILS_RETURN
} from 'constants/testIds'

export const DetailsPokemon = () => {
  const {
    name,
    details,
    sprites,
    evolutions,
    speciesName,
    isErrorDetails,
    isFetchingDetails,
    isFetchingEvolutions
  } = usePokemonDetails()
  const navigate = useNavigate()

  const handleGoBack = () => {
    navigate('/')
  }

  if (isFetchingDetails) {
    return <div data-testid={TESTID_DETAILS_LOADING}>Loading...</div>
  }

  if (isErrorDetails) {
    return <div data-testid={TESTID_DETAILS_ERROR}>Invalid Pokemon.</div>
  }

  return (
    <>
      <Box display="flex" mt={2} data-testid={TESTID_DETAILS_POKEMON}>
        <Box flexShrink={0} position="relative">
          <IconButton
            onClick={handleGoBack}
            data-testid={TESTID_DETAILS_RETURN}
            sx={{ position: 'absolute', top: 0, left: 0 }}
          >
            <ArrowBackIcon />
          </IconButton>
          <img
            alt={name}
            src={sprites?.other?.['official-artwork']?.front_default ?? ''}
          />
        </Box>
        <Box p={2} flexGrow={1}>
          <Box mb={2}>
            <Typography component="h1" variant="h4">
              {name}
            </Typography>
            <Typography fontSize={14} color="grey.700" sx={{ mt: 1 }}>
              species: {speciesName}
            </Typography>
          </Box>
          {details.map(({ title, content }) => (
            <Card
              key={`details-${title}`}
              sx={{ bgcolor: 'secondary.main', color: 'common.white', mb: 2 }}
            >
              <CardHeader title={title} />
              <CardContent>{content}</CardContent>
            </Card>
          ))}
        </Box>
      </Box>
      <GridEvolution
        evolutions={evolutions}
        isFetching={isFetchingEvolutions}
      />
    </>
  )
}
