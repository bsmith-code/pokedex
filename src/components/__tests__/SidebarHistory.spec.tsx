import { useMemo } from 'react'
import { mockSearchHistory } from '__mocks__/searchHistory'
import { fireEvent, screen, within } from '@testing-library/react'

import { IPokemonContext, PokemonContext } from 'context/PokemonContext'

import { SidebarHistory } from 'components/SidebarHistory'

import { TESTID_SEARCH_EMPTY, TESTID_SEARCH_HISTORY } from 'constants/testIds'

const defaultValue = {
  searchHistory: mockSearchHistory,
  handleChangePokemon: jest.fn()
}

const MockComponent = (overrideValue: Partial<IPokemonContext>) => {
  const value = useMemo(() => ({ ...defaultValue, ...overrideValue }), [])

  return (
    <PokemonContext.Provider value={value as IPokemonContext}>
      <SidebarHistory />
    </PokemonContext.Provider>
  )
}

describe('SidebarHistory', () => {
  it('renders correctly WITHOUT data', () => {
    const { container } = customRender(<MockComponent searchHistory={[]} />)

    expect(screen.getByTestId(TESTID_SEARCH_EMPTY)).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('renders correctly WITH data', () => {
    const { container } = customRender(<MockComponent />)

    expect(screen.getByTestId(TESTID_SEARCH_HISTORY)).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('correctly changes pokemon', () => {
    customRender(<MockComponent />)

    const searchHistory = screen.getByTestId(TESTID_SEARCH_HISTORY)
    const searchItem = within(searchHistory).getAllByRole('button')[4]
    fireEvent.click(searchItem)

    const searchValue = searchItem.textContent ?? ''
    expect(defaultValue.handleChangePokemon).toHaveBeenCalledWith(searchValue)
  })
})
