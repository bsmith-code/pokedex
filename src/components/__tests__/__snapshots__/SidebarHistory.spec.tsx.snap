// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarHistory renders correctly WITH data 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiBox-root emotion-0"
  >
    <h6
      class="MuiTypography-root MuiTypography-subtitle2 emotion-1"
    >
      Search History
    </h6>
    <ul
      class="MuiList-root MuiList-padding emotion-2"
      data-testid="search-history"
    >
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 0
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 1
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 2
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 3
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 4
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 5
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 6
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 7
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 8
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 9
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 10
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 11
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 12
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 13
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 14
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 15
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 16
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 17
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 18
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-3"
        role="button"
        tabindex="0"
      >
        Search 19
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </li>
    </ul>
  </div>
</div>
`;

exports[`SidebarHistory renders correctly WITHOUT data 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiBox-root emotion-0"
  >
    <h6
      class="MuiTypography-root MuiTypography-subtitle2 emotion-1"
    >
      Search History
    </h6>
    <p
      class="MuiTypography-root MuiTypography-body1 emotion-2"
      data-testid="search-empty"
    >
      No previous searches.
    </p>
  </div>
</div>
`;
